frame/staking-rewards/src/lib.rs

```diff
@@ -457,59 +539,159 @@ pub mod pallet {
 														)?,
 													)
 													.map_err(|_| ArithmeticError::Overflow)?;
+												if reward_asset == nft.asset {
+													PendingTotalShares::<T>::mutate(
+														nft.asset,
+														|total_shares| {
+															let reward_shares: u128 = reward_shares;
+															*total_shares += reward_shares;
+														},
+													);
+												}
 											}
 										},
 									}
 									Ok(())
 								},
 							);
+
 							if let Err(e) = try_reward {
-								log::warn!("Failed to reward NFT: {:?}, message: {:?}", nft_id, e);
+								log::error!("Failed to reward NFT: {:?}, message: {:?}", nft_id, e);
 							}
 						},
 					);
 					if let BlockFold::Done { .. } = result {
-						CurrentState::<T>::set(State::Registering);
+						CurrentState::<T>::set(State::PendingAmounts);
 					}
 				},
-				State::Registering => {
-					// TODO: extend adding extensions to exisitng nfts
-					let result = <(FoldState<T>, PendingStakers<T>)>::step(
-						FoldStrategy::Chunk {
-							number_of_elements: T::ElementToProcessPerBlock::get(),
-						},
-						(),
-						|_, nft_id, _| {
-							let nft = T::get_protocol_nft::<StakingNFTOf<T>>(&nft_id)
-								.expect("impossible; qed");
-							for reward_asset in nft.pending_rewards.keys() {
-								TotalShares::<T>::mutate(
-									(nft.asset, reward_asset),
-									|total_shares| {
-										*total_shares = total_shares
-											.checked_add(nft.shares())
-											.expect("impossible; qed;");
-									},
-								);
-							}
-							Stakers::<T>::insert(nft_id, ());
-						},
-					);
-					if let BlockFold::Done { .. } = result {
-						PendingStakers::<T>::remove_all(None);
-						CurrentState::<T>::set(State::WaitingForEpochEnd);
+				State::PendingAmounts => {
+					for (nft_id, amount) in PendingAmountExtensions::<T>::drain()
+						.take(T::ElementToProcessPerBlock::get() as usize)
+					{
+						any = true;
+						let result = T::try_mutate_protocol_nft(
+							&nft_id,
+							|nft: &mut StakingNFTOf<T>| -> DispatchResult {
+								let old_shares = nft.shares();
```

Comment:
Note: With the new proposed design this code block would not be necessary as new shares are accounted for by the pool by having reward inflation for each rewarded asset without having to wait for the epoch to finish.